SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialMetricsDisplay.cpp

```diff
@@ -0,0 +1,187 @@
+#include "Engine/Canvas.h"
+#include "EngineClasses/SpatialNetDriver.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Net/UnrealNetwork.h"
+#include "Utils/SpatialMetrics.h"
+
+ASpatialMetricsDisplay::ASpatialMetricsDisplay(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+{
+	PrimaryActorTick.bCanEverTick = true;
+	PrimaryActorTick.bStartWithTickEnabled = true;
+
+	bReplicates = true;
+	bAlwaysRelevant = true;
+
+	NetUpdateFrequency = 1.f;
+}
+
+void ASpatialMetricsDisplay::BeginPlay()
+{
+	Super::BeginPlay();
+
+	WorkerStats.Reserve(PreallocatedWorkerCount);
+	WorkerStatsLastUpdateTime.Reserve(PreallocatedWorkerCount);
+}
+
+void ASpatialMetricsDisplay::Destroyed()
+{
+	Super::Destroyed();
+
+	if (DrawDebugDelegateHandle.IsValid())
+	{
+		UDebugDrawService::Unregister(DrawDebugDelegateHandle);
+	}
+}
+
+bool ASpatialMetricsDisplay::ServerUpdateWorkerStats_Validate(const float Time, const FWorkerStats& OneWorkerStats)
+{
+	return true;
+}
+
+void ASpatialMetricsDisplay::ServerUpdateWorkerStats_Implementation(const float Time, const FWorkerStats& OneWorkerStats)
+{
+	int32 StatsIndex = WorkerStats.Find(OneWorkerStats);
+
+	if (StatsIndex == INDEX_NONE)
+	{
+		StatsIndex = WorkerStats.AddDefaulted();
+	}
+
+	WorkerStats[StatsIndex] = OneWorkerStats;
+
+	float& WorkerUpdateTime = WorkerStatsLastUpdateTime.FindOrAdd(OneWorkerStats.WorkerName);
+	WorkerUpdateTime = Time;
+}
+
+void ASpatialMetricsDisplay::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(ASpatialMetricsDisplay, WorkerStats);
+}
+
+void ASpatialMetricsDisplay::DrawDebug(class UCanvas* Canvas, APlayerController* Controller)
+{
+	enum StatColumns
+	{
+		StatColumn_Worker,
+		StatColumn_AverageFPS,
+		StatColumn_WorkerLoad,
+		StatColumn_Last
+	};
+
+	const uint32 StatDisplayStartX = 25;
+	const uint32 StatDisplayStartY = 80;
+
+	const FString StatColumnTitles[StatColumn_Last] = { TEXT("Worker"), TEXT("FPS"), TEXT("Load") };
+	const uint32 StatColumnOffsets[StatColumn_Last] = { 0, 160, 80 };
+	const uint32 StatRowOffset = 20;
+
+	const FString StatSectionTitle = TEXT("Spatial Metrics Display");
+
+	if (GetWorld()->IsServer())
+	{
+		return;
+	}
+
+	FFontRenderInfo FontRenderInfo = Canvas->CreateFontRenderInfo(false, true);
+
+	UFont* RenderFont = GEngine->GetSmallFont();
+
+	uint32 DrawX = StatDisplayStartX;
+	uint32 DrawY = StatDisplayStartY;
+
+	Canvas->SetDrawColor(FColor::Green);
+	Canvas->DrawText(RenderFont, StatSectionTitle, DrawX, DrawY, 1.0f, 1.0f, FontRenderInfo);
+	DrawY += StatRowOffset;
+
+	Canvas->SetDrawColor(FColor::White);
+	for (int i = 0; i < StatColumn_Last; ++i)
+	{
+		DrawX += StatColumnOffsets[i];
+		Canvas->DrawText(RenderFont, StatColumnTitles[i], DrawX, DrawY, 1.0f, 1.0f, FontRenderInfo);
+	}
+
+	DrawY += StatRowOffset;
+
+	for (const FWorkerStats& OneWorkerStats : WorkerStats)
```

Comment:
Do we not need to increase `DrawY` in this loop?

Reply:
yes good catch ieva.  I haven't successfully tested against multiserver yet as it crashes on NWX.